choral/src/main/java/choral/compiler/amend/RelaxedTyper.java

```diff
@@ -0,0 +1,2828 @@
+				for( VariableDeclaration x : n.variables() ) {
+					GroundDataType type = visitGroundDataTypeExpression( scope, x.type(), false );
+					scope.declareVariable( x.name().identifier(), type );
+					x.initializer().ifPresent( e -> synth( scope, e, method ) );
+				}
+				return assertReachableContinuation( n, false );
+			}
+
+			public boolean assertReachableContinuation( Statement n, boolean returnChecked ) {
+				if( returnChecked && n.hasContinuation() ) {
+					throw StaticVerificationException.of(
+							"unreachable statement",
+							n.continuation().position() );
+				}
+				returnChecked |= visit( n.continuation() );
+				n.setReturnAnnotation( returnChecked );
+				return returnChecked;
+			}
+		}
+
+		/**
+		 * A relaxed version of Synth. Doesn't throw exceptions on data location mismatch.
+		 */
+		private final class Synth extends AbstractChoralVisitor< GroundDataTypeOrVoid > {
+
+			public Synth( VariableDeclarationScope scope, HigherMethod method ) {
+				this( scope, false, method );
+			}
+
+			public Synth( 
+				VariableDeclarationScope scope, 
+				boolean explicitConstructorArg, 
+				HigherMethod method 
+			) {
+				this.scope = scope;
+				this.explicitConstructorArg = explicitConstructorArg;
+				this.method = method;
+			}
+
+			public Synth( 
+				VariableDeclarationScope scope, 
+				List< ? extends World > homeWorlds, 
+				HigherMethod method   
+			) {
+				this( scope, false, homeWorlds, method );
+			}
+
+			public Synth( 
+				VariableDeclarationScope scope, boolean explicitConstructorArg, 
+				List< ? extends World > homeWorlds,
+				HigherMethod method 
+			) {
+				this.scope = scope;
+				this.explicitConstructorArg = explicitConstructorArg;
+				this.homeWorlds = homeWorlds;
+				this.method = method;
+			}
+			
+			private final VariableDeclarationScope scope;
+			private GroundDataTypeOrVoid left = null;
+			private boolean leftStatic = false;
+			private final boolean explicitConstructorArg;
+			List< ? extends World > homeWorlds = Collections.emptyList(); // The worlds at which the expression takes place
+			
+			boolean checkLocation = true; 	// for scopedexpressions like "this.obj.val" we only want to check location once for 
+											// the whole expression, not once for every sub-expression
+			String fullName;				// the full name of a scopedExpression, used by innermost scopedExpression
+			HigherMethod method;			// the enclosing method
+
+			@Override
+			public GroundDataTypeOrVoid visit( Expression n ) {
+				return n.accept( this );
+			}
+
+			@Override
+			public GroundDataTypeOrVoid visit( ScopedExpression n ) {
+				// We want to capture the full ScopedExpression, but only consieder the worldargument 
+				// of the innermost expression 
+				// for expression obj.first.second.val, we want to capture the full expression, but 
+				// only consider the world of val
+
+				if(checkLocation){ // only true at the first visited ScopedExpression in a scoped chain
```

Comment:
```suggestion
				// Consider a program like:
				// ```
				// MyObj@B obj = ...;
				// int@A x = obj.first.second + 1@A;
				// ```
				// We want to infer that B needs to send `obj.first.second` to A. This means:
				// 1. When we get to the root of the scoped expression, record its full name.
				// 2. We disable communication inference for all sub-expressions of the scoped
				//    expression, except the innermost one.

				if(checkLocation){
```

Reply:
I found your explanation a little too technical to understand. Take a look at this version and add anything you think is wrong or missing